wpilibNewCommands/src/test/native/cpp/frc2/command/AddRequirementsTest.cpp

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+#include <wpi/array.h>
+
+#include "CommandTestBase.h"
+#include "frc2/command/Command.h"
+#include "frc2/command/RunCommand.h"
+
+using namespace frc2;
+
+class OverloadResolver {
+ public:
+  OverloadResolver(std::function<void()> requirementsOverload,
+                   std::function<void()> smallSetOverload,
+                   std::function<void()> subsystemOverload)
+      : m_requirementsOverload{std::move(requirementsOverload)},
+        m_smallSetOverload{std::move(smallSetOverload)},
+        m_subsystemOverload{std::move(subsystemOverload)} {}
+
+  void AddRequirements(Requirements requirements) { m_requirementsOverload(); }
+
+  void AddRequirements(wpi::SmallSet<Subsystem*, 4> requirements) {
+    m_smallSetOverload();
+  }
+
+  void AddRequirements(Subsystem* requirement) { m_subsystemOverload(); }
+
+ private:
+  std::function<void()> m_requirementsOverload;
+  std::function<void()> m_smallSetOverload;
+  std::function<void()> m_subsystemOverload;
+};
+
+TEST(AddRequirementsTest, OverloadResolution) {
+  TestSubsystem requirement;
+
+  int requirementsOverloadCounter = 0;
+  int smallSetOverloadCounter = 0;
+  int subsystemOverloadCounter = 0;
+
+  OverloadResolver overloadResolver([&] { requirementsOverloadCounter++; },
+                                    [&] { smallSetOverloadCounter++; },
+                                    [&] { subsystemOverloadCounter++; });
+
+  // Check initializer list
+  // Have to use two elements to select the Requirements overload instead of the
+  // Subsystem* overload
+  overloadResolver.AddRequirements({&requirement, &requirement});
+  EXPECT_EQ(requirementsOverloadCounter, 1);
+  EXPECT_EQ(smallSetOverloadCounter, 0);
+  EXPECT_EQ(subsystemOverloadCounter, 0);
+
+  // Check span
+  std::span<Subsystem* const> requirementsSpan;
+
+  overloadResolver.AddRequirements(requirementsSpan);
+  EXPECT_EQ(requirementsOverloadCounter, 2);
+  EXPECT_EQ(smallSetOverloadCounter, 0);
+  EXPECT_EQ(subsystemOverloadCounter, 0);
+
+  // Check wpi::SmallSet
+  wpi::SmallSet<Subsystem*, 4> requirementsSet;
+
+  overloadResolver.AddRequirements(requirementsSet);
+  EXPECT_EQ(requirementsOverloadCounter, 2);
+  EXPECT_EQ(smallSetOverloadCounter, 1);
+  EXPECT_EQ(subsystemOverloadCounter, 0);
+
+  // Check subsystem pointer
+  overloadResolver.AddRequirements(&requirement);
+  EXPECT_EQ(requirementsOverloadCounter, 2);
+  EXPECT_EQ(smallSetOverloadCounter, 1);
+  EXPECT_EQ(subsystemOverloadCounter, 1);
+}
+
+TEST(AddRequirementsTest, InitializerListSemantics) {
+  TestSubsystem requirement1;
+  TestSubsystem requirement2;
+
+  RunCommand command([] {});
+  command.AddRequirements({&requirement1, &requirement2});
```

Comment:
Is there a reason in all these that the requirements are added in the following line rather than inline?

Reply:
I wanted to specifically test `AddRequirements()` (which actually isn't used in the tests except for `SelfCancellingCommand` in `SchedulingRecursionTest`). Maybe those tests should also be in Java, though, since it's not tested there either?

Reply:
If I remember correctly, `RunCommand` etc call `AddRequirements` with the `Requirements` object they're given. So it would check everything other than directly calling `AddRequirements` -- so this is correct. Never mind.